book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -0,0 +1,305 @@
+that's required to process some request asynchronously is to clone the
+appropriate handle, move it into an async block, and make the call as part of
+the future. (We might want to use Tokio's blocking API for this, but this is
+an implementation detail).
+
+Because the state service has exclusive access to the sled database, and the
+state service sees a linearized sequence of state requests, we have an easy
+way to opt in to asynchronous database access. We can perform sled operations
+synchronously in the `Service::call`, waiting for them to complete, and be
+sure that all future requests will see the resulting sled state. Or, we can
+perform sled operations asynchronously in the future returned by
+`Service::call`.
+
+If we perform all *writes* synchronously and allow reads to be either
+synchronous or asynchronous, we ensure that writes cannot race each other.
+Asynchronous reads are guaranteed to read at least the state present at the
+time the request was processed, or a later state.
+
+In summary:
+
+- **Sled reads** may be done synchronously (in `call`) or asynchronously (in
+  the `Future`), depending on the context;
+
+- **Sled writes** must be done synchronously (in `call`).
+
+## In-memory data structures
+[in-memory]: #in-memory
+
+At a high level, the in-memory data structures store a collection of chains,
+each rooted at the highest finalized block. Each chain consists of a map from
+heights to blocks. Chains are stored using an ordered map from difficulty to
+chains, so that the map ordering is the ordering of best to worst chains.
+
+- XXX fill in details on exact types
+
+- XXX work out whether we should store extra data (e.g., a HashSet of UTXOs
+  spent by some block etc.) to speed up checks.
+
+When a new block extends the best chain past 100 blocks, the old root is
+removed from the in-memory state and committed to sled.
+
+## Sled data structures
+[sled]: #sled
+
+Sled provides a persistent, thread-safe `BTreeMap<&[u8], &[u8]>`. Each map is
+a distinct "tree". Keys are sorted using lex order on byte strings, so
+integer values should be stored using big-endian encoding (so that the lex
+order on byte strings is the numeric ordering).
+
+We use the following Sled trees:
+
+| Tree                 |                  Keys |                              Values |
+|----------------------|-----------------------|-------------------------------------|
+| `blocks_by_hash`     | `block::Hash`     | `Block`                             |
+| `hash_by_height`     | `BE32(height)`        | `block::Hash`                   |
+| `tx_by_hash`         | `transaction::Hash`     | `block::Hash || BE32(tx_index)` |
+| `utxo_by_outpoint`   | `OutPoint`            | `TransparentOutput`                 |
+| `sprout_nullifiers`  | `sprout::Nullifier`   | `()`                                |
+| `sapling_nullifiers` | `sapling::Nullifier`  | `()`                                |
```

Comment:
```suggestion
| Tree                 |                  Keys |                              Values |
|----------------------|-----------------------|-------------------------------------|
| `blocks_by_hash`     | `block::Hash`     | `Block`                             |
| `hash_by_height`     | `BE32(height)`        | `block::Hash`                   |
| `tx_by_hash`         | `transaction::Hash`     | `block::Hash || BE32(tx_index)` |
| `utxo_by_outpoint`   | `OutPoint`            | `TransparentOutput`                 |
| `sprout_nullifiers`  | `sprout::Nullifier`   | `()`                                |
| `sapling_nullifiers` | `sapling::Nullifier`  | `()`                                |
| `sprout_anchors`  | `sprout::tree::Root`   | `()`                                |
| `sapling_anchors` | `sapling::tree::Root`  | `()`                                |
```

I don't _think_ we need to be able to look up anchors by height. 

For `JoinSplit`s, they may include an anchor that refers to another `JoinSplit` in the same transaction, how should we handle those?

Reply:
Yeah I think looking up anchors by height is error-prone, because there can be multiple blocks at the same height

Reply:
In this case, there's no risk of error, because we consider all the Sled data to be finalized. So there cannot be multiple blocks at the same height, at least for blocks stored in Sled.

Reply:
Ah right, that's what I was missing.